### ISSUE TITLE
Simulate Blocking with logging [Feature Request]

### ISSUE THREAD

**Issue by soumsps** [CIDRAM/CIDRAM#98]
Hello @Maikuolan,
I was looking for a way to monitor my traffic .... but sometimes I just want to monitor instead of blocking/showing captcha.

So, according to me it would be nice feature if we could have **Simulate Blocking** feature in CIDRAM. In this mode CIDRAM will log all request ... that will be blocked/shown captcha ...   But will not block in real... it will just log.

**Comment by DanielRuf** [comment]
CIDRAM is not meant for monitoring the (real) traffic which is more than just http as protocol (ftp, ssh, smtp, tls, pop3/imap, dns, ...).

**Comment by soumsps** [comment]
No, I am not saying monitor all protocol... just monitor and log those traffic which should be blocked by CIDRAM rule ....  BUT Don't block 

I wish there could be some feature that could tell me what amount of traffic is blockable .. according to CIDRAM rules ....

**Comment by ghost** [comment]
Exactly how is that supposed to work?

A non-working installation?
Taking into account only installed modules or all available modules, or just specific ones?

Personally I'm with @danielruf on this one, CIDRAM is not a monitoring tool. Scope creep warning in being.

If you want to check some data, you can grab the server log, filter out the data you want and run it through the front end test page manually.

**Comment by DanielRuf** [comment]
> If you want to check some data, you can grab the server log, filter out the data you want and run it through the front end test page manually.

Exactly what I would do, replay(ing) requests.

**Comment by Maikuolan** [comment]
Although CIDRAM itself is not a monitoring tool, it could (theoretically) be used as part of a monitoring tool by leveraging the external __[CIDRAM API](https://github.com/CIDRAM/CIDRAM-Extras/tree/master/api)__ (assuming, of course, that you want to test whether actual live, real, inbound requests would normally be blocked by CIDRAM, as opposed to just checking specific, pre-determined IPs or user agents). No need really to build new features in order for this to work. :-)

I need to update the external CIDRAM API to be able to also work with the new auxiliary rules feature (I'd actually kind of forgotten about the external CIDRAM API when I first implemented auxiliary rules to the CIDRAM core, so, auxiliary rules support for the external CIDRAM API is something I still need to do; shouldn't take too long to do that though). Otherwise though, it should still all be synced and up-to-date with the CIDRAM core and all features available to the CIDRAM core otherwise.

If you've already got the external CIDRAM API up and running somewhere, you could do something like this to set it up as a monitoring tool, if you wanted (just scribbled this up just now in Notepad++ as an example and to get in a quick reply; I haven't thoroughly tested it; may need to play around with it a bit anyway, to get it working exactly as needed):

```PHP
<?php
require __DIR__ . '/api.php'; // ..or the /path/to/the/api.php file

$Foo = new \CIDRAM\API\API($CIDRAM);

$Results = $Foo->lookup(
    $_SERVER[$CIDRAM['IPAddr']],
    true,
    empty($_SERVER['HTTP_USER_AGENT']) ? '' : $_SERVER['HTTP_USER_AGENT']
);

if ($Results['SignatureCount']) {
    file_put_contents(
        '/path/to/a/file/somewhere/to/log/monitored/stuff.txt',
        sprintf(
            "Simulated block: %s - %s - %s\n",
            time(),
            $_SERVER[$CIDRAM['IPAddr']],
            empty($_SERVER['HTTP_USER_AGENT']) ? '' : $_SERVER['HTTP_USER_AGENT']
        ),
        FILE_APPEND
    );
}
```

If you're only wanting to test specific theoretical requests though (e.g., whether specific IPs or user agents would normally be blocked by CIDRAM), as opposed to actual live requests, then the front-end IP test page would be the best way to go.

**Comment by Maikuolan** [comment]
Otherwise though, if we were to implement such a feature (or if someone wanted to hack it in themselves), the only part of the CIDRAM core that we/they would really need to change (from a functional perspective), is the final `die` in the output generator:

```PHP
    /** All necessary processing and logging has completed; Now we send HTML output and die. */
    die($CIDRAM['HTML']);
```

Basically, either commenting out that `die` completely (the hack approach, to permanently disable until uncommented again), or wrapping it in an `if()` statement (the implementation approach, to enable/disable it optionally) to execute the `die` only if a specific configuration directive is enabled (the other considerations for an implementation approach being, of course, updating the configuration directives YAML, the corresponding documentation, and the corresponding L10N, so that it appears on the configuration page, and so that people will know what it is and does via the documentation; basically the same routine whenever any new configuration directives are added).

**Comment by Maikuolan** [comment]
Anyway, I'm going to get some sleep for the night (pretty late where I am at the moment, and getting a bit tired). I'll think on it a bit and try to reply a little better when I'm a bit more awake.

**Comment by DanielRuf** [comment]
If, if should be probably an extension. But in general: at least changes needed to make it possible in some way but not as core feature (imho).

**Comment by soumsps** [comment]
> Exactly how is that supposed to work?

To be exact, I need no change in the core , I was in need of just a mode, which can monitor bad traffic even when CIDRAM is not active...

> A non-working installation?
> Taking into account only installed modules or all available modules, or just specific ones?

Ummm, I just wanted to take into account for all selected signatures in configuration.. :)
but, allowing specific one is also good ... but I guess may there could be performance issue ... if a user selected many modules.

> /** All necessary processing and logging has completed; Now we send HTML output and die. */
>     die($CIDRAM['HTML']);

Thank you for this hack @Maikuolan 

This is what i was asking for ... although when simply commented this die() function. then for one visitor multiple log ID was generated ... But on later i got to know its due to some AJAX request and admin-ajax.php

With Auxillary Rules, I have whitelisted these AJAX reconstructed URI. So right now my CIDRAM is on Dry Run(Simulate Mode) .. :) 

![screenshot_8](https://user-images.githubusercontent.com/17058535/48044198-bf9c8600-e1b0-11e8-9d7e-0af227d1a7ff.jpg)

_I completely understand that CIDRAM is not a monitoring tool, and we can surely omit this as a feature._

@Maikuolan 
Still I would ask for simple switch in Maintenance mode (True/False/ Dry Run mode). With this little warning on top.
![screenshot_9](https://user-images.githubusercontent.com/17058535/48044399-c5469b80-e1b1-11e8-8fae-a0655c8e1e72.jpg)

Anyways, I got my required solution, and I can change `outgen.php` time to time whenever required :)

Thank You

**Comment by Maikuolan** [comment]
Adding labels *suggestion* and *planning*.

Will look at this again properly after I'm back home again from my overseas trip at the end of this month (I don't anticipate implementing any new features, significant changes, or releasing any new versions of any packages - CIDRAM, phpMussel, etc - until that time at the earliest; better to give such things the dedicated time that such things deserve, rather than rushing it while on the move, I think).

**Comment by Maikuolan** [comment]
*Related: #221.*

Also.. "8 Nov 2018". Been a while, aye..? My bad. "o_o

I'll get this sorted out soon.

**Comment by Maikuolan** [comment]
Implemented for v3 (8d5f2a3fea197611f6cde32c25ba5aa0843ab06b). :-)

![dry](https://user-images.githubusercontent.com/12571108/158070288-0f5051f0-bd10-4e22-b2b0-b2eec48419e7.png)
